gnoi/os/client.go

```diff
@@ -27,11 +30,170 @@ import (
 // Client handles requesting OS RPCs.
 type Client struct {
 	client pb.OSClient
+	read   func(string) (file io.ReaderAt, size uint64, close func() error, err error)
 }
 
+const chunkSize = 5000000
+
 // NewClient returns a new OS service client.
 func NewClient(c *grpc.ClientConn) *Client {
-	return &Client{client: pb.NewOSClient(c)}
+	reader := func(path string) (file io.ReaderAt, size uint64, close func() error, err error) {
+		f, err := os.Open(path)
+		if err != nil {
+			return
+		}
+		fileInfo, err := f.Stat()
+		if err != nil {
+			return
+		}
+		size = uint64(fileInfo.Size())
+		file = f
+		close = f.Close
+		return
+	}
+	return &Client{client: pb.NewOSClient(c), read: reader}
+}
+
+// Install invokes the Install RPC for the OS service.
+func (c *Client) Install(ctx context.Context, imgPath, version string, printStatus bool, validateTimeout time.Duration) error {
+	// Open and Stat OS image.
+	if c.read == nil {
+		return fmt.Errorf("No reader passed to client")
+	}
+	file, fileSize, fileClose, err := c.read(imgPath)
+	if err != nil {
+		return err
+	}
+	defer fileClose()
+
+	// Create Install client for streaming.
+	install, err := c.client.Install(ctx)
```

Comment:
repackage this context with a cancel such that you can cancel it and avoid blocked Send/Recv:
https://godoc.org/context#WithCancel